lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/ManagedStrategyBase.kt

```diff
@@ -0,0 +1,485 @@
+/*-
+ * #%L
+ * Lincheck
+ * %%
+ * Copyright (C) 2019 JetBrains s.r.o.
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.jetbrains.kotlinx.lincheck.strategy
+
+import org.jetbrains.kotlinx.lincheck.*
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionResult
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario
+import org.jetbrains.kotlinx.lincheck.util.PseudoRandom
+import org.jetbrains.kotlinx.lincheck.verifier.Verifier
+import java.lang.IllegalStateException
+import java.util.*
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.IdentityHashMap
+
+abstract class ManagedStrategyBase(
+        testClass: Class<*>,
+        scenario: ExecutionScenario,
+        verifier: Verifier,
+        reporter: Reporter,
+        maxRepetitions: Int,
+        protected val requireObstructionFreedom: Boolean
+) : ManagedStrategy(testClass, scenario, verifier, reporter) {
+    protected val parallelActors: List<List<Actor>> = scenario.parallelExecution
+
+    // whether a thread finished all its operations
+    protected val finished: Array<AtomicBoolean> = Array(nThreads) { AtomicBoolean(false) }
+    // what thread is currently allowed to perform operations
+    @Volatile
+    protected var currentThread: Int = 0
+    // detector of loops (i.e. active locks)
+    protected val loopDetector = LoopDetector(maxRepetitions)
+    // logger of all events in the execution such as thread switches
+    protected var eventLogger: ThreadEventLogger = DummyEventLogger
+    // tracker of acquisitions and releases of monitors
+    protected val monitorTracker = MonitorTracker(nThreads)
+    // random used for execution
+    protected var executionRandom = PseudoRandom()
+    // copy of random used for execution for reproducability of results
+    protected var executionRandomCopy = executionRandom
```

Comment:
Can't you just introduce a 'lastValue()' method? AFAIR, you already keep these values.